tensorflow/core/kernels/dynamic_stitch_op.cc

```diff
@@ -20,9 +20,119 @@ limitations under the License.
+      auto indices_vec = indices.flat<int32>();
+      const Tensor& data = data_inputs[input_num];
+      auto data_flat =
+          data.shaped<T, 2>({indices_vec.dimension(0), slice_size});
+
+      if (DataTypeCanUseMemcpy(DataTypeToEnum<T>::v())) {
+        T* merged_base = &merged_flat(0, 0);
+        const T* data_base = &data_flat(0, 0);
+        const size_t slice_bytes = slice_size * sizeof(T);
+        for (int i = 0; i < indices_vec.size(); i++) {
+          int32 index = internal::SubtleMustCopy(indices_vec(i));
+          OP_REQUIRES(
+              c, FastBoundsCheck(index, first_dim_size),
+              errors::InvalidArgument("indices[", i, "] is out of range"));
+          memcpy(merged_base + index * slice_size, data_base + i * slice_size,
+                 slice_bytes);
+        }
+      } else {
+        Eigen::DSizes<Eigen::DenseIndex, 2> sizes(1, slice_size);
+        for (int i = 0; i < indices_vec.size(); i++) {
+          // Copy slice data[i] to merged[indices[i]]
+          Eigen::DSizes<Eigen::DenseIndex, 2> data_indices(i, 0);
+          int32 index = internal::SubtleMustCopy(indices_vec(i));
+          OP_REQUIRES(
+              c, FastBoundsCheck(index, first_dim_size),
+              errors::InvalidArgument("indices[", i, "] is out of range"));
+          Eigen::DSizes<Eigen::DenseIndex, 2> merged_indices(index, 0);
+          merged_flat.slice(merged_indices, sizes) =
+              data_flat.slice(data_indices, sizes);
+        }
+      }
+    }
+  }
+};
+
+#if GOOGLE_CUDA
+
+template <typename T>
+struct DynamicStitchGPULaunch {
+  static void Run(const GPUDevice& d, const int slice_size,
+                  const int first_dim_size,
+                  const CudaDeviceArrayStruct<int32>& indices_flat,
+                  const CudaDeviceArrayStruct<const T*>& data_slice_ptrs,
+                  T* output);
+};
+
+template <typename T>
+struct LaunchDynamicStitchOp<GPUDevice, T> {
+  static void launch(OpKernelContext* c, const int slice_size,
+                     const int first_dim_size, OpInputList indices_inputs,
+                     OpInputList data_inputs, Tensor* merged) {
+    // createt two arrays that will be sent to CUDA device
+    // one used as pointers to output's row id
+    CudaDeviceArrayOnHost<int32> indices_flat(c, first_dim_size);
+    // another used as pointers to the pointers of the head of each data_slice
+    CudaDeviceArrayOnHost<const T*> data_slice_ptrs(c, first_dim_size);
+    OP_REQUIRES_OK(c, indices_flat.Init());
+    OP_REQUIRES_OK(c, data_slice_ptrs.Init());
+
+    int ptr_num = 0;
+    for (int input_num = 0; input_num < indices_inputs.size(); input_num++) {
+      auto indices_vec = indices_inputs[input_num].flat<int32>();
+      auto base_ptr = data_inputs[input_num].template flat<T>().data();
+      for (int ind_num = 0; ind_num < indices_vec.size(); ind_num++) {
+        indices_flat.Set(ptr_num, indices_vec(ind_num));
```

Comment:
No, that is a bad way to handle collisions.  It doesn't respect the documented semantics at all: the order of writes would be entirely random.  Also, it doesn't work if each output index corresponds to a slice with more than one element, and would require very slow atomic operations.

A better way would be to have a single array indexed by `first_dim_size` that's computed on the CPU with correct collision resolution.  In the case of no uninitialized entries in the output array (the common case), this array will be no larger than current arrays.

However, this whole approach of doing a large chunk of the work on CPU seems low quality at best.  I don't see any way around it given the collision requirements, though, and the current fake GPU implementation is probably strictly worse.  Unpleasant!